packages/opentelemetry-exporter-collector/src/transform.ts

```diff
@@ -0,0 +1,241 @@
+import { Attributes, Link, TimedEvent, TraceState } from '@opentelemetry/types';
+import * as collectorTypes from './types';
+
+const OT_MAX_STRING_LENGTH = 128;
+const OT_MAX_ATTRIBUTES = 30;
+
+const LINK_TYPE_UNSPECIFIED: collectorTypes.LinkTypeUnspecified = 0;
+// const LINK_TYPE_CHILD_LINKED_SPAN: collectorTypes.LinkTypeChildLinkedSpan = 1;
+const LINK_TYPE_PARENT_LINKED_SPAN: collectorTypes.LinkTypeParentLinkedSpan = 2;
+
+/**
+ * convert string to maximum length of 128, providing information of truncated bytes
+ * @param name - string to be converted
+ */
+export function toCollectorTruncatableString(
+  name: string
+): collectorTypes.TruncatableString {
+  const value = name.substr(0, OT_MAX_STRING_LENGTH);
+  const truncatedByteCount =
+    name.length > OT_MAX_STRING_LENGTH ? name.length - OT_MAX_STRING_LENGTH : 0;
+
+  return { value, truncatedByteCount };
+}
+
+/**
+ * convert attributes
+ * @param attributes
+ * @param maxAttributes - default value is 30
+ */
+export function toCollectorAttributes(
+  attributes: Attributes,
+  maxAttributes: number = OT_MAX_ATTRIBUTES
+): collectorTypes.Attributes {
+  const attributeMap: collectorTypes.AttributeMap = {};
+  let droppedAttributesCount = 0;
+
+  const keys = Object.keys(attributes || {});
+
+  const countKeys = Math.min(keys.length, maxAttributes);
+  if (keys.length > maxAttributes) {
+    droppedAttributesCount = keys.length - maxAttributes;
+  }
+
+  for (let i = 0; i <= countKeys - 1; i++) {
+    const key = keys[i];
+    const eventAttributeValue = toCollectorEventValue(
+      attributes && attributes[key]
+    );
+    if (eventAttributeValue) {
+      attributeMap[key] = eventAttributeValue;
+    }
+  }
+
+  return {
+    droppedAttributesCount,
+    attributeMap,
+  };
+}
+
+/**
+ * convert event value
+ * @param value event value
+ */
+export function toCollectorEventValue(
+  value: unknown
+): collectorTypes.AttributeValue | undefined {
+  const attributeValue: collectorTypes.AttributeValue = {};
+
+  if (typeof value === 'string') {
+    attributeValue.stringValue = toCollectorTruncatableString(value);
+  } else if (typeof value === 'boolean') {
+    attributeValue.boolValue = value;
+  } else if (typeof value === 'number') {
+    if (Math.floor(value) === value) {
+      attributeValue.intValue = value;
+    } else {
+      attributeValue.doubleValue = value;
+    }
+  }
+
+  return attributeValue;
+}
+
+/**
+ * convert events
+ * @param events array of events
+ * @param maxAttributes - maximum number of event attributes to be converted
+ */
+export function toCollectorEvents(
+  events: TimedEvent[],
+  maxAttributes: number = OT_MAX_ATTRIBUTES
+): collectorTypes.TimeEvents {
+  let droppedAnnotationsCount = 0;
+  let droppedMessageEventsCount = 0; // not counting yet as messageEvent is not implemented
+
+  const timeEvent: collectorTypes.TimeEvent[] = events.map(
+    (event: TimedEvent) => {
+      let attributes: collectorTypes.Attributes | undefined;
```

Comment:
Optional/nit: can you do `let attributes?: collectorTypes.Attributes;`?

Reply:
hmm from what I see you cannot define variable as optional, this is the interface / function param thing. Typescript is complaining about such structure `let attributes?: collectorTypes.Attributes;`

Reply:
OK, good to know!